src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -310,55 +313,71 @@ public ShardRouting routingEntry() {
         return this.shardRouting;
     }
 
-    public IndexShard routingEntry(ShardRouting newRouting) {
-        ShardRouting currentRouting = this.shardRouting;
+    /**
+     * Updates the shards routing entry. This mutate the shards internal state depending
+     * on the changes that get introduced by the new routing value. This method will persist shard level metadata
+     * unless explicitly disabled.
+     */
+    public void updateRoutingEntry(final ShardRouting newRouting, final boolean persistState) {
+        final ShardRouting currentRouting = this.shardRouting;
         if (!newRouting.shardId().equals(shardId())) {
             throw new ElasticsearchIllegalArgumentException("Trying to set a routing entry with shardId [" + newRouting.shardId() + "] on a shard with shardId [" + shardId() + "]");
         }
-        if (currentRouting != null) {
-            if (!newRouting.primary() && currentRouting.primary()) {
-                logger.warn("suspect illegal state: trying to move shard from primary mode to replica mode");
-            }
-            // if its the same routing, return
-            if (currentRouting.equals(newRouting)) {
-                return this;
+        try {
+            if (currentRouting != null) {
+                assert newRouting.version() > currentRouting.version() : "expected: " + newRouting.version() + " > " + currentRouting.version();
```

Comment:
I see now it was in the original code as well. I think it's worth testing we never go back *on the same master*, but perhaps we should do that in the cluster state management part, in updateTask, before publishing the CS?